### PR TITLE
TINY-7373: Prevent editing of content whilst progress state is true

### PR DESCRIPTION
**Pull request by ltrouton**
Related Ticket: TINY-7373

Description of Changes:
* Update the Throbber.ts to take focus when it is enabled and return focus back to the editor when disabled

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [X] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        focus,
+        getRoot: () => memBlockRoot.getOpt(component),
+      })
+    ])
+  });
+
+  return component;
+};
+
+describe('browser.alloy.blocking.BlockingFocusTest', () => {
```
Comment (lnewson) [inline modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts:53]:
```suggestion
describe('browser.alloy.behaviour.blocking.BlockingFocusTest', () => {
```

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts [hunk 2]
```diff
@@ -11,16 +11,19 @@ export type BlockFn = (
   state: BlockingState,
   getBusySpec: (root: AlloyComponent, behaviour: Behaviour.AlloyBehaviourRecord) => AlloySpec
 ) => void;
+
 export type UnblockFn = (component: AlloyComponent, config: BlockingConfig, state: BlockingState) => void;
 
 export interface BlockingConfigSpec extends Behaviour.BehaviourConfigSpec {
   readonly getRoot?: (component: AlloyComponent) => Optional<AlloyComponent>;
+  readonly focus?: boolean;
```
Comment (lnewson) [inline modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts:19]: Alloy changelog entry for this please

- [ ] Add alloy changelog entry

File: modules/tinymce/src/core/main/ts/focus/EditorFocus.ts [hunk 3]
```diff
@@ -119,6 +119,11 @@ const focusEditor = (editor: Editor) => {
     normalizeSelection(editor, rng);
   }
 
+  // This should be safe as the the focusin listener calls editor.focus(true) which means the focusEditor function is not called
+  // TODO: Should this also include the last blurredEditor?
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/focus/EditorFocus.ts:123]: There's a couple of TODOs here. Are they still accurate? If so could we please make sure to either, fix, log a JIRA or remove them.

File: modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts [hunk 4]
```diff
@@ -62,19 +54,59 @@ const renderThrobber = (spec): AlloySpec => ({
...
+    })
   ]),
   components: [ ]
 });
 
+const focusBusyComponent = (throbber: AlloyComponent): void => {
+  Composing.getCurrent(throbber).each((comp) => {
+    Keying.focusIn(comp);
+  });
+};
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts:72]:
```suggestion
const focusBusyComponent = (throbber: AlloyComponent): void =>
  Composing.getCurrent(throbber).each(Keying.focusIn);
```
This can be ETA reduced and can be flattened

File: modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts [hunk 5]
```diff
@@ -62,19 +54,59 @@ const renderThrobber = (spec): AlloySpec => ({
...
+    if (throbberState.get()) {
+      focusBusyComponent(lazyThrobber());
+    }
+  });
+
   const toggle = (state: boolean) => {
     if (state !== throbberState.get()) {
-      toggleThrobber(lazyThrobber(), state, sharedBackstage.providers);
+      const throbber = lazyThrobber();
+      const throbberFocus = Composing.getCurrent(throbber).exists(Focusing.isFocused);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts:106]: This isn't used in all paths, so I think this would be cleaner if you only did the lookup inside `handleFocus` for the `state === false` path.

- [ ] Move this inside the `handleFocus` function

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    base_url: '/project/tinymce/js/tinymce',
+  }, [ Theme ]);
+
+  const setProgressState = (editor: Editor, state: boolean, time?: number) => {
+    if (state) {
+      editor.setProgressState(true, time);
+    } else {
+      editor.setProgressState(false);
+    }
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts:19]:
```suggestion
    editor.setProgressState(state, time);
```
This is overly complicated and can be slimmed down to the above. Realistically you probably don't even need this function and could call `setProgressState` directly.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+          type: 'submit',
+          name: 'save',
+          text: 'Save',
+          primary: true
+        }
+      ],
+    });
+  };
+
+  const setProgressState = (editor: Editor, state: boolean, time?: number) => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:54]: See comments on the previous test. This is also overly complicated and can be simplified.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    if (state) {
+      editor.setProgressState(true, time);
+    } else {
+      editor.setProgressState(false);
+    }
+  };
+
+  const pAssertThrobberVisible = async () => {
+    await UiFinder.pWaitForVisible('Wait for throbber to show', SugarBody.body(), '.tox-throbber');
+  };
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:64]:
```suggestion
  const pAssertThrobberVisible = () =>
    UiFinder.pWaitForVisible('Wait for throbber to show', SugarBody.body(), '.tox-throbber');
```
Remember we prefer single-line lambdas/arrow functions where possible. There's also no need for that async/await here.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+  const pAssertThrobberHidden = async () => {
+    await UiFinder.pWaitForHidden('Wait for throbber to hide', SugarBody.body(), '.tox-throbber');
+  };
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:68]:
```suggestion
  const pAssertThrobberHidden = () =>
    UiFinder.pWaitForHidden('Wait for throbber to hide', SugarBody.body(), '.tox-throbber');
```

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+  const pAssertThrobberHasFocus = async () => {
+    await FocusTools.pTryOnSelector('Throbber has focus', SugarDocument.getDocument(), 'div.tox-throbber__busy-spinner');
+  };
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:72]:
```suggestion
  const pAssertThrobberHasFocus = () =>
    FocusTools.pTryOnSelector('Throbber has focus', SugarDocument.getDocument(), 'div.tox-throbber__busy-spinner');
```

File: modules/tinymce/src/core/main/ts/focus/EditorFocus.ts [hunk 11]
```diff
@@ -119,6 +119,11 @@ const focusEditor = (editor: Editor) => {
...
+  // TODO: Need some tests for this
+  editor.fire('focus');
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/focus/EditorFocus.ts:125]: Actually, why was this added? This almost definitely seems incorrect as we'll end up with the focus event firing twice.

**Review by lnewson — CHANGES_REQUESTED** [review]
Overall looks good, but marking as needs changes because you've added the `editor.fire(focus)` part which we definitely cannot do so we'll need another solution there.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  };
+
+  it('TINY-7373: should not change the editor selection', async () => {
+    const editor = hook.editor();
+    editor.setContent('<p>abc</p>');
+    TinySelections.setSelection(editor, [ 0, 0 ], 0, [ 0, 0 ], 2);
+    setProgressState(editor, true);
+    await UiFinder.pWaitForVisible('Wait for throbber to show', SugarBody.body(), '.tox-throbber');
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts:27]: Could these wait for's be pulled out into a separate function to remove the duplication in each test? You could possibly even pull out the whole lot. eg:
```
const pTestToggle = async (editor: Editor, action: () => void = Fun.noop) => {
  setProgressState(editor, true);
  await UiFinder.pWaitForVisible('Wait for throbber to show', SugarBody.body(), '.tox-throbber');
  action();
  setProgressState(editor, false);
  await UiFinder.pWaitForHidden('Wait for throbber to hide', SugarBody.body(), '.tox-throbber');
}
```

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts (hunk 2 above)
Comment (ltrouton) [reply modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts:19]: Thanks. Forgot about that

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts (hunk 6 above)
Comment (ltrouton) [reply modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts]: Thanks. Have fixed this across all of the test files

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts (hunk 12 above)
Comment (ltrouton) [reply modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberEditorTest.ts]: Thanks for the suggestion. Looks a lot cleaner now.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+    // Throbber should get focus when notification is closed instead of editor
+    notification.close();
+    await Waiter.pTryUntil('Notification should close', () => UiFinder.notExists(SugarBody.body(), 'div.tox-notification'));
+    await pAssertThrobberFocus();
+
+    // Focus returns to the editor when throbber is closed (this should happen since a dialog will normally focus the editor on close)
+    await pDisableThrobber(editor);
+    assert.isTrue(editor.hasFocus());
+  });
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:174]: - [ ] Can you add some tests to make sure that users clicking on / tabbing into the editor doesn't work while the throbber is up

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/focus/EditorFocus.ts (hunk 3 above)
Comment (ltrouton) [reply modules/tinymce/src/core/main/ts/focus/EditorFocus.ts]: Have removed the code and TODOs here

File: modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts (hunk 5 above)
Comment (ltrouton) [reply modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts]: Have moved the focus logic into the `toggleThrobber` function as it is simple enough now and for toggle off, I need check the throbber focus before it is removed.

File: modules/tinymce/src/core/main/ts/focus/EditorFocus.ts (hunk 11 above)
Comment (ltrouton) [reply modules/tinymce/src/core/main/ts/focus/EditorFocus.ts]: Have added an `EditorFocus` event that is fired on `editor.focus()` since listening to `focus` is not sufficient to detect all focuses of the editor.

File: modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts (hunk 2 above)
Comment (lnewson) [reply modules/alloy/src/main/ts/ephox/alloy/behaviour/blocking/BlockingTypes.ts:19]: Marking as resolved, but don't forget to fix the missing line Neil mentioned.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    await pAssertThrobberFocus();
+    assert.isFalse(editor.hasFocus());
+
+    await pDisableThrobber(editor);
+    assert.isTrue(editor.hasFocus());
+  });
+
+  it('TINY-7373: should have correct focus transitions when opening and closing dialog', async () => {
+    const editor = hook.editor();
+    const pAssertDialogFocus = () => FocusTools.pTryOnSelector('Dialog input has focus (1)', SugarDocument.getDocument(), 'input.tox-textfield');
```
[the remark this answers is not in the document]
Comment (ltrouton) [reply modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts]: Was leftover so thanks for noting that.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts (hunk 13 above)
Comment (ltrouton) [reply modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:174]: Thanks for the suggestion. I have added the tabbing test to make sure that works correctly. I think a clicking test would be better served as a manual test as it should not be possible for the user to click the editor since the throbber div covers the whole editor. In order to test clicking would require testing a scenario that user couldn't actually do themselves.

File: modules/tinymce/src/core/main/ts/api/EditorCommands.ts [hunk 15]
```diff
@@ -289,6 +292,10 @@ class EditorCommands {
         editor.undoManager.add();
       },
 
+      'mceFocus': (_command, _ui, value?: boolean) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/EditorCommands.ts:295]: Added changelog entry for this new command maybe?

File: modules/tinymce/src/core/main/ts/undo/Setup.ts [hunk 16]
```diff
@@ -12,6 +12,18 @@ import * as Levels from './Levels';
 import { endTyping, setTyping } from './TypingState';
 import { Locks, UndoLevel, UndoManager } from './UndoManagerTypes';
 
+const skipCommand = (cmd: string): boolean => {
+  switch (cmd) {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/undo/Setup.ts:16]:
```suggestion
  switch (cmd.toLowerCase()) {
```
This would be cleaner and would allow us to remove the `toLowerCase()` calls on lines 42 and 52.

File: modules/tinymce/src/core/main/ts/undo/Setup.ts [hunk 17]
```diff
@@ -12,6 +12,18 @@ import * as Levels from './Levels';
 import { endTyping, setTyping } from './TypingState';
 import { Locks, UndoLevel, UndoManager } from './UndoManagerTypes';
 
+const skipCommand = (cmd: string): boolean => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/undo/Setup.ts:15]:
```suggestion
const ignoreCommand = (cmd: string): boolean => {
```
This would probably be more accurate as to what it's doing. Another option would be `shouldIgnoreCommand`.

File: modules/tinymce/src/core/test/ts/browser/UndoManagerTest.ts [hunk 18]
```diff
@@ -556,4 +556,11 @@ describe('browser.tinymce.core.UndoManager', () => {
     editor.execCommand('mceRepaint');
     assert.isFalse(editor.undoManager.hasUndo());
   });
+
+  it('undo filter for mceFocus is case insensitive', () => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/UndoManagerTest.ts:560]: JIRA number prefix?

File: modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts [hunk 19]
```diff
@@ -25,27 +27,54 @@ const renderSpinner = (providerBackstage: UiFactoryBackstageProviders): AlloySpe
...
+// When the throbber is enabled, prevent the iframe from being part of the sequential keyboard navigation when Tabbing
+// TODO: TINY-7500 Only works for iframe mode at this stage
+const toggleEditorTabIndex = (editor: Editor, state: boolean) => {
+  Optional.from(editor.iframeElement)
+    .map(SugarElement.fromDom)
+    .each((iframe) => {
+      if (state) {
+        Attribute.set(iframe, 'tabindex', -1);
+      } else {
+        Obj.get(Settings.getIframeAttrs(editor), 'tabindex')
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/throbber/Throbber.ts:44]: I'm not sure this is the best way to go about this... while it will restore the original value, what if someone had changed it after initialization? It might be better to store the original as a data attribute and restore using that instead.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+    TinyUiActions.cancelDialog(editor);
+    await Waiter.pTryUntil('Dialog should close', () => UiFinder.notExists(SugarBody.body(), 'div[role="dialog"]'));
+    await pAssertThrobberFocus();
+
+    // Focus returns to the editor when throbber is closed (this should happen since a dialog will normally focus the editor on close)
+    await pDisableThrobber(editor, pAssertEditorFocus(editor));
+    assert.isTrue(editor.hasFocus());
+  });
+
+  // In theory, this combination shouldn't be possible as the user cannot click any buttons to open dialogs when the throbber is enabled
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberFocusTest.ts:150]: This is definitely a viable case, as take for example a notification could be shown (as it is with RTC) while the throbber is active.

File: modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberTest.ts [hunk 21]
```diff
@@ -1,5 +1,5 @@
 import { ApproxStructure, Assertions, UiFinder, Waiter } from '@ephox/agar';
-import { describe, it } from '@ephox/bedrock-client';
+import { assert, describe, it } from '@ephox/bedrock-client';
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/throbber/ThrobberTest.ts:2]: Don't use the legacy `assert` module, use chai instead please.

- [ ] Use chai assertions

File: modules/tinymce/src/themes/silver/test/ts/webdriver/throbber/ThrobberTabbingTest.ts [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+
+    await pPressTab('div.tox-throbber__busy-spinner');
+    await pAssertInputFocus(false)();
+
+    await pDisableThrobber(editor, pAssertInputFocus(false));
+    assert.isFalse(editor.hasFocus());
+  });
+
+  // Automated test doesn't work on Safari but works manually
+  if (!Env.browser.isSafari()) {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/throbber/ThrobberTabbingTest.ts:96]: This isn't how we skip tests, it should be done inside the function instead. eg:
```
it('TINY-7373: should be able to Shift+Tab out of the throbber if it has focus', async function () {
  if (Env.browser.isSafari()) {
    this.skip();
  }
```

- [ ] Fix how this test is being skipped on Safari

**Review by lnewson — APPROVED** [review]

File: modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+      }));
+      await FocusTools.pTryOnSelector('Busy component has focus', SugarDocument.getDocument(), 'div.put-spinner-here');
+      Blocking.unblock(comp);
+    });
+
+    it('does not focus if busy component does not have Keying configured', async () => {
+      const comp = hook.component();
+      Focusing.focus(memFocusDiv.get(comp));
+      Blocking.block(comp, Fun.constant({ dom: { tag: 'div', classes: [ 'put-spinner-here' ] }}));
+      await FocusTools.pTryOnSelector('Busy component has focus', SugarDocument.getDocument(), 'div.focus-div');
```
Comment (JamesToohey) [inline modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts:75]: Should this be
```suggestion
      await FocusTools.pTryOnSelector('Busy component does not have focus', SugarDocument.getDocument(), 'div.focus-div');
```

File: modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+      const comp = hook.component();
+      Focusing.focus(memFocusDiv.get(comp));
+      Blocking.block(comp, (_comp, behaviours) => ({
+        dom: {
+          tag: 'div',
+          classes: [ 'put-spinner-here' ]
+        },
+        behaviours
+      }));
+      await FocusTools.pTryOnSelector('Busy component has focus', SugarDocument.getDocument(), 'div.focus-div');
```
Comment (JamesToohey) [inline modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts:93]: Same here?
```suggestion
      await FocusTools.pTryOnSelector('Busy component does not have focus', SugarDocument.getDocument(), 'div.focus-div');
```

Comment (JamesToohey) [reply modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts:93]: I wonder if these might be clearer if the classes/selectors for the busy component and the div that should be focused are pulled out into constants?

It might be nicer to do a `:not()` selector too?

File: modules/tinymce/src/core/main/ts/undo/Setup.ts [hunk 25]
```diff
@@ -12,6 +12,18 @@ import * as Levels from './Levels';
 import { endTyping, setTyping } from './TypingState';
 import { Locks, UndoLevel, UndoManager } from './UndoManagerTypes';
 
+const shouldIgnoreCommand = (cmd: string): boolean => {
+  switch (cmd.toLowerCase()) {
```
Comment (JamesToohey) [inline modules/tinymce/src/core/main/ts/undo/Setup.ts:16]: Could we add a comment explaining why these commands should be ignored?

File: modules/tinymce/CHANGELOG.md [hunk 26]
```diff
@@ -8,11 +8,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - Added a new `toolbar_sticky_offset` setting to allow the toolbar to be offset from the top of the page #TINY-7337
+- Added a new `mceFocus` command that focuses the editor. Equivalent to using `editor.focus()` #TINY-7373
 
 ### Changed
 - Changed the load order so that the content css gets loaded before the editor gets populated with contents. #TINY-7249
 
 ### Fixed
+- The editor content could be edited after calling `setProgressState(true)` #TINY-7373
```
Comment (JamesToohey) [inline modules/tinymce/CHANGELOG.md:17]: Do we need to mention that this fix only applies to iframe mode?

**Review by JamesToohey — APPROVED** [review]

File: modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts (hunk 23 above)
Comment (ltrouton) [reply modules/alloy/src/test/ts/browser/behaviour/blocking/BlockingFocusTest.ts]: Thanks. Have fixed and refactored this.

File: modules/tinymce/CHANGELOG.md (hunk 26 above)
Comment (ltrouton) [reply modules/tinymce/CHANGELOG.md]: Good point. Have added `...in iframe mode` to the end but not quite sure if that is clear enough

**Review by ashfordneil — APPROVED** [review]
